server/handlers/project_command_output_handler.go

```diff
@@ -85,47 +94,58 @@ func (p *DefaultProjectCommandOutputHandler) Clear(ctx models.ProjectCommandCont
 }
 
 func (p *DefaultProjectCommandOutputHandler) clearLogLines(pull string) {
-	p.controllerBufferLock.Lock()
+	p.projectOutputBuffersLock.Lock()
 	delete(p.projectOutputBuffers, pull)
-	p.controllerBufferLock.Unlock()
+	p.projectOutputBuffersLock.Unlock()
 }
 
-func (p *DefaultProjectCommandOutputHandler) addChan(pull string) chan string {
-	ch := make(chan string, 1000)
-	p.controllerBufferLock.Lock()
-	for _, line := range p.projectOutputBuffers[pull] {
-		ch <- line
-	}
+func (p *DefaultProjectCommandOutputHandler) addChan(ch chan string, pull string) {
+	p.receiverBuffersLock.Lock()
 	if p.receiverBuffers[pull] == nil {
 		p.receiverBuffers[pull] = map[chan string]bool{}
 	}
 	p.receiverBuffers[pull][ch] = true
-	p.controllerBufferLock.Unlock()
-	return ch
+	p.receiverBuffersLock.Unlock()
```

Comment:
There are so many locks being used everywhere it makes everything kind of difficult to understand/read. Not necessary to change as part of this PR, but as a thought exercise it would be good to think about basically having a single thread manage receiver buffers and projectBuffers and basically using channels to communicate between that thread and other threads thereby eliminating the need to have locks everywhere.

Reply:
I wonder if we can implement something more event driven? We can register new webhooks to output_handler instances and emit any new requests to that webhook that should remove the need to using channels to sending messages to the webhook sessions.

Reply:
Hmm that's true. I was trying to fix what was already working but I could definitely work on some refactoring and making the code-flow more understandable. I'll work on it after this PR.